request_reader.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"encoding/json"
+	"io"
+
+	gyaml "github.com/ghodss/yaml"
+	"gopkg.in/yaml.v2"
+)
+
+type requestReader interface {
+	next() ([]byte, error)
+}
+
+// jsonRequestReader parses multiple JSON objects from given reader
+// JSON objects can be delimited by space or newline
+type jsonRequestReader struct{ *json.Decoder }
+
+func (r *jsonRequestReader) next() ([]byte, error) {
+	if !r.More() {
+		return nil, io.EOF
+	}
+	var v json.RawMessage
+	err := r.Decode(&v)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(v), nil
+}
+
+// yamlRequestReader parses multiple YAML objects from given reader
+// YAML objects must be delimited by `---`
+type yamlRequestReader struct{ *yaml.Decoder }
+
+// next decodes the request from the reader as YAML or JSON and returns
+// JSON encoded byte
+func (r *yamlRequestReader) next() ([]byte, error) {
+	var v interface{}
+	err := r.Decode(&v)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := yaml.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	return gyaml.YAMLToJSON(bytes)
```

Comment:
Are you sure we need to do this? Both, the Thrift and Proto serializers expect YAML input. (JSON is valid YAML so JSON works too, but in this case, just the raw YAML would work.)
```suggestion
	return bytes
```

Reply:
True, this would work in the current state. I wanted to make the contract in serializer to expect JSON input only to avoid redundant (un)marshals. This has been mostly addressed in the refactored branch